src/crimson/os/seastore/extent_placement_manager.cc

```diff
@@ -0,0 +1,135 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab expandtab
+
+#include "crimson/os/seastore/extent_placement_manager.h"
+#include "crimson/os/seastore/lba_manager/btree/lba_btree_node_impl.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.h"
+#include "crimson/os/seastore/onode_manager/staged-fltree/node_extent_manager/seastore.h"
+#include "crimson/os/seastore/collection_manager/collection_flat_node.h"
+#include "crimson/os/seastore/object_data_handler.h"
+
+namespace crimson::os::seastore {
+
+ExtentAllocWriter::write_ertr::future<>
+ExtentAllocWriter::write(CachedExtent* extent) {
+  bufferlist bl;
+  extent->prepare_write();
+  bl.append(extent->get_bptr());
+  auto segment = extent->get_allocated_segment();
+  assert(extent->extent_writer);
+  assert(segment);
+  auto old_segment_iter = open_segments.begin();
+  if (segment.get() != old_segment_iter->get()) {
+    // segment writes are ordered, so when an opened segment isn't the current
+    // one being written to, it is full and should be closed.
+    (void) open_segments.front()->close();
+    open_segments.erase(old_segment_iter);
+  }
+  return segment->write(extent->get_paddr().offset, bl).safe_then(
+    [this, extent, segment] {
+    extent->extent_writer = nullptr;
+    extent->clear_allocated_segment();
+  });
+}
+
+ExtentAllocWriter::alloc_extent_ertr::future<alloc_t>
+ExtentAllocWriter::alloc(segment_off_t length) {
+  auto roll_fut = roll_segment_ertr::make_ready_future<>();
+  if (_needs_roll(length)) {
+    roll_fut = roll_segment();
+  }
+  return roll_fut.safe_then([this, length] {
+    paddr_t addr{current_segment->get_segment_id(), allocated_to};
+    alloc_t alloc_addr{addr, current_segment};
+    allocated_to += length;
+    return alloc_addr;
+  });
```

Comment:
I think each type device will have its own alloc algorithm.  This part code is a wrapper for each device, right? am I misunderstanding something?